core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java

```diff
@@ -844,19 +845,29 @@ public List<ECKey> getImportedKeys() {
     }
 
     /**
-     * @return Long.MAX_VALUE if empty
+     * Gets the earliest time for which full block must be downloaded.
+     * @return earliest creation times of keys in this group,
+     *         {@link Instant#EPOCH} if at least one time is unknown,
+     *         {@link Instant#MAX} if no keys in this group
      */
+    public Instant getEarliestKeyCreationTimeInstant() {
+        return TimeUtils.earlier(basic.getEarliestKeyCreationTimeInstant(), getEarliestChainsCreationTime());
+    }
+
+    /** @deprecated use {@link #getEarliestKeyCreationTimeInstant()} */
+    @Deprecated
     public long getEarliestKeyCreationTime() {
-        return Math.min(basic.getEarliestKeyCreationTime(), getEarliestChainsCreationTime());
+        Instant earliestKeyCreationTime = getEarliestKeyCreationTimeInstant();
+        return earliestKeyCreationTime.equals(Instant.MAX) ? Long.MAX_VALUE : earliestKeyCreationTime.getEpochSecond();
     }
 
-    private long getEarliestChainsCreationTime() {
+    private Instant getEarliestChainsCreationTime() {
         if (chains == null)
-            return Long.MAX_VALUE;
+            return Instant.MAX;
         return chains.stream()
-                .mapToLong(DeterministicKeyChain::getEarliestKeyCreationTime)
-                .min()
-                .orElse(Long.MAX_VALUE);
+                .map(DeterministicKeyChain::getEarliestKeyCreationTimeInstant)
+                .min((t1, t2) -> t1.compareTo(t2))
+                .orElse(Instant.MAX);
```

Comment:
Ternary operator.

Reply:
Github is truncating the code too much and I was unclear. I mean the entire method:

```
return (chains == null)
             ? Instant.MAX;
             : chains.stream()
                 .map(DeterministicKeyChain::getEarliestKeyCreationTimeInstant)
                 .min((t1, t2) -> t1.compareTo(t2))
                 .orElse(Instant.MAX);
```

Reply:
Ok thanks for clarifying. This is again a case of easier review vs getting it right in the first place. Since this is a larger refactoring, I'm leaning towards easier review. But I'll put a commit on top so that we don't forget.

Reply:
Done: a0a6a116b09f2235eeebb2482011f1eb9f2c76e2